cardano-testnet/src/Testnet/Components/Query.hs

```diff
@@ -98,25 +106,111 @@ waitForEpochs epochStateView@EpochStateView{nodeConfigPath, socketPath} interval
+  void . liftIO . atomically $ do
+    -- Drain the channel first, to not keep anything not used in memory. We only need to store one element in
+    -- the channel. This is a safeguard against elements piling up in the channel If there are no listeners.
+    _ <- runMaybeT . forever . MaybeT $ tryReadTChan wakeLock
+    -- notify all listeners on duplicated channels
+    writeTChan wakeLock ()
+
+-- | Watch epoch state view for an update. On every update, the callback function gets executed.
+watchEpochStateUpdate
+  :: HasCallStack
+  => MonadTest m
+  => MonadIO m
+  => EpochStateView
+  -> ((AnyNewEpochState, SlotNo, BlockNo) -> m (Maybe a))
+  -- ^ callback function invoked repeatedly. Stops the processing when 'Just a' gets returned
+  -> m a -- ^ the result from the callback function
+watchEpochStateUpdate EpochStateView{wakeLock, epochStateView} f = withFrozenCallStack $
+  -- dupTChan for multi-wakeup
+  go =<< (liftIO . atomically $ dupTChan wakeLock)
+    where
+      go wakeupChannel = do
+        newEpochState <- liftIO . atomically $ do
+          _ <- readTChan wakeupChannel -- block and wait for update
+          readTMVar epochStateView
+        f newEpochState >>= \case
+          Nothing -> go wakeupChannel
+          Just a -> pure a
+
 -- | Get epoch state from the view. If the state isn't available, retry waiting up to 15 seconds. Fails when
 -- the state is not available after 15 seconds.
-getEpochState :: MonadTest m
-              => MonadAssertion m
-              => MonadIO m
-              => EpochStateView
-              -> m AnyNewEpochState
+getEpochState
+  :: HasCallStack
+  => MonadTest m
+  => MonadAssertion m
+  => MonadIO m
+  => EpochStateView
+  -> m AnyNewEpochState
 getEpochState EpochStateView{epochStateView} =
   withFrozenCallStack $
     H.byDurationM 0.5 15 "EpochStateView has not been initialized within 15 seconds" $
-      H.evalIO (readIORef epochStateView) >>= maybe H.failure pure
+      H.evalIO (atomically $ tryReadTMVar epochStateView) >>= maybe H.failure (\(nes, _, _) -> pure nes)
+
+getBlockNumber
+  :: HasCallStack
+  => MonadIO m
+  => MonadTest m
+  => MonadAssertion m
+  => EpochStateView
+  -> m BlockNo -- ^ The number of last produced block
+getBlockNumber EpochStateView{epochStateView} =
+  withFrozenCallStack $
+    H.byDurationM 0.5 15 "EpochStateView has not been initialized within 15 seconds" $
+      H.evalIO (atomically $ tryReadTMVar epochStateView) >>= maybe H.failure (\(_, _, blockNumber) -> pure blockNumber)
 
+getSlotNumber
+  :: HasCallStack
+  => MonadIO m
+  => MonadTest m
+  => MonadAssertion m
+  => EpochStateView
+  -> m SlotNo -- ^ The current slot number
+getSlotNumber EpochStateView{epochStateView} =
+  withFrozenCallStack $
+    H.byDurationM 0.5 15 "EpochStateView has not been initialized within 15 seconds" $
+      H.evalIO (atomically $ tryReadTMVar epochStateView) >>= maybe H.failure (\(_, slotNumber, _) -> pure slotNumber)
```

Comment:
`getEpochState`, `getBlockNumber`, and `getSlotNumber` are identical except for the lambda. We could have an auxiliary function (which doesn't need to be exported) that is used by all three

Reply:
Good idea, I'll do that.